include/zephyr/device.h

```diff
@@ -1075,24 +1075,16 @@ device_get_dt_nodelabels(const struct device *dev)
 		Z_DEVICE_SECTION_NAME(level, prio), DEVICE_NAME_GET(dev_id)) =                     \
 		Z_DEVICE_INIT(name, pm, data, config, api, state, deps, node_id, dev_id)
 
-/* deprecated device initialization levels */
-#define Z_DEVICE_LEVEL_DEPRECATED_EARLY                                        \
-	__WARN("EARLY device driver level is deprecated")
-#define Z_DEVICE_LEVEL_DEPRECATED_PRE_KERNEL_1
-#define Z_DEVICE_LEVEL_DEPRECATED_PRE_KERNEL_2
-#define Z_DEVICE_LEVEL_DEPRECATED_POST_KERNEL
-#define Z_DEVICE_LEVEL_DEPRECATED_APPLICATION                                  \
-	__WARN("APPLICATION device driver level is deprecated")
-#define Z_DEVICE_LEVEL_DEPRECATED_SMP                                          \
-	__WARN("SMP device driver level is deprecated")
-
 /**
- * @brief Issue a warning if the given init level is deprecated.
+ * @brief Issue an error if the given init level is not supported.
  *
  * @param level Init level
  */
-#define Z_DEVICE_LEVEL_CHECK_DEPRECATED_LEVEL(level)                           \
-	Z_DEVICE_LEVEL_DEPRECATED_##level
+#define Z_DEVICE_CHECK_INIT_LEVEL(level)                                       \
+	COND_CODE_1(Z_INIT_PRE_KERNEL_1_##level, (),                           \
+	(COND_CODE_1(Z_INIT_PRE_KERNEL_2_##level, (),                          \
+	(COND_CODE_1(Z_INIT_POST_KERNEL_##level, (),                           \
+	(ZERO_OR_COMPILE_ERROR(0)))))))
```

Comment:
Does it make sense to do something like this:

```suggestion
	(ZERO_OR_COMPILE_ERROR("ERROR: given init level is not supported")))))))
```

so that the error message is less cryptic?

```
In file included from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util.h:17,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/atomic.h:18,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/kernel_includes.h:25,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /Users/ycsin/zephyrproject/zephyr/drivers/sensor/blah/blah.c:3:
/Users/ycsin/zephyrproject/zephyr/include/zephyr/device.h:1087:11: error: expected identifier or '(' before string constant
 1087 |         (("ERROR: given init level is not supported")))))))
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

current:

```
In file included from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util_macro.h:34,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util.h:17,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/atomic.h:18,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/kernel_includes.h:25,
                 from /Users/ycsin/zephyrproject/zephyr/include/zephyr/kernel.h:17,
                 from /Users/ycsin/zephyrproject/zephyr/drivers/sensor/ams/blah/blah.c:3:
/Users/ycsin/zephyrproject/zephyr/include/zephyr/sys/util.h:97:39: error: expected identifier or '(' before 'int'
   97 | #define ZERO_OR_COMPILE_ERROR(cond) ((int) sizeof(char[1 - 2 * !(cond)]) - 1)
      |                                       ^~~
```